### PR TITLE
Add PointerEvent.endOfBatch

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR adds a new field `bool endOfBatch` to `PointerEvent` and populates it.

Some pointer events are sent in batches. For example, when the device is being operated with multiple fingers, events from each moving finger constitute a batch and is sent together. This information is necessary to implement the iOS behavior of [multi-finger scrolling](https://github.com/flutter/flutter/pull/106552), which needs to know which event is the last of the batch so that it can start processing the entire batch as a whole.

**What constitutes a batch?** The iOS platform sends a batch of events in the same [`allTouches`](https://developer.apple.com/documentation/uikit/uievent/1613836-alltouches?language=objc), and the iOS embedding is already sending events from the same `allTouches` in the same `PointerDataPacket`. The framework just needs to mark all of them as `endOfBatch: false` except for the last one.

The harder part is when the events are being resampled. If there are multiple batches within two resampling frames, the two batches might not have the same order, or even the same set of devices. This PR considers all events resulting from traversing `PointerEventResampler`s and calling `PointerEventResampler.sample()` a single batch. All of these events are collected, and are marked as `endOfBatch: false` except for the last one.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by goderbauer** [comment]
Does this play nicely/correctly when pointer events are resampled with [PointerEventResampler](https://master-api.flutter.dev/flutter/gestures/PointerEventResampler-class.html)? Maybe add a test for that case?

**Comment by dkwingsmt** [comment]
> Does this play nicely/correctly when pointer events are resampled with [PointerEventResampler](https://master-api.flutter.dev/flutter/gestures/PointerEventResampler-class.html)? Maybe add a test for that case?

That's a good point and I suspect that I do need extra changes.... I'll take a look soon.

**Comment by goderbauer** [comment]
I have this on my list to review early next week. Just a quick fly-by comment, looks like the doc check is unhappy.

File: packages/flutter/lib/src/gestures/resampler.dart [hunk 1]
```diff
@@ -245,16 +253,21 @@ class PointerEventResampler {
         // therefore never produce `hover` events.
         if (position != _position) {
           final Offset delta = position - _position;
-          callback(_toMoveOrHoverEvent(event, position, delta,
-              _pointerIdentifier, sampleTime, wasDown, hadButtons));
+          eventBuilders.add((bool endOfBatch) =>
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/resampler.dart:256]: nit (here and elsewhere): https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#consider-using--for-short-functions-and-methods:

> [...] only use ⇒ when everything, including the function declaration, fits on a single line.

File: packages/flutter/lib/src/gestures/resampler.dart [hunk 2]
```diff
@@ -272,8 +285,10 @@ class PointerEventResampler {
     final PointerEvent? next = _next;
     if (position != _position && next != null) {
       final Offset delta = position - _position;
-      callback(_toMoveOrHoverEvent(next, position, delta, _pointerIdentifier,
-          sampleTime, _isDown, _hasButtons));
+      eventBuilders.add((bool endOfBatch) =>
+        _toMoveOrHoverEvent(next, position, delta, _pointerIdentifier,
+            sampleTime, _isDown, _hasButtons, endOfBatch),
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/resampler.dart:290]: nit: formatting

File: packages/flutter/lib/src/gestures/resampler.dart [hunk 3]
```diff
@@ -335,3 +354,54 @@ class PointerEventResampler {
...
+  ///
+  /// The `nextSampleTime` is the timestamp of the next frame. A positive value
+  /// allows early processing of up and removed events. This improves
+  /// resampling of these events, which is important for fling animations.
+  void sample(Duration sampleTime, Duration nextSampleTime, HandleEventCallback handlePointerEvent) {
+    // Iterate over active resamplers and sample pointer events for
+    // current sample time.
+    final List<BatchEventBuilder> batchBuilder = <BatchEventBuilder>[];
+    for (final PointerEventResampler resampler in _resamplers.values) {
+      batchBuilder.addAll(resampler.sample(sampleTime, nextSampleTime));
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/resampler.dart:383]: While reading this I have been wondering why we need all this indirection via the `BatchEventBuilder`. For my own understanding: Is it possible that `resampler.sample` here my return an empty list and the `endOfBatch` event is actually produced by a Resampler other then the last one in `_resamplers.values`? If that's not true, could we just tell the last `resampler.sample` method that it is responsible for producing an `endOfBatch` event?

Comment (goderbauer) [reply packages/flutter/lib/src/gestures/resampler.dart:383]: If we have to sample all resamplers before we can determine the endOfBatch event, could this be simplified by just having resampler.sample return a list of pointer events and before calling handlePointerEvent below we copy the last event of the collected list to set `endOfBatch: true`?

**Comment by goderbauer** [comment]
Does this actually help with solving https://github.com/flutter/flutter/pull/106552 in all cases? Let's imagine a situation where the screen is split in two halves. One half is covered by a DragGestureRecognizer, the other half isn't. Two finger (F1 and F2) touch the screen and move around, one in each half. F1 moves around in the half covered by the DragGestureRecognizer, which tracks its every move. F2 moves around outside of that area and is not tracked by the DragGestureRecognizer. In a single PointerData packet (`Iterable<ui.PointerData>`) we now get a move event for F1 and another move event for F2, which are processed by the `PointerEventConverter` in that order. The F2 event is marked with `endOfBatch`, the F1 event isn't. Unless I am missing something, the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it... Wouldn't this cause problems for https://github.com/flutter/flutter/pull/106552 (I haven't reviewed the latest changes to that PR closely, this is just from conceptually thinking about this)?

**Comment by dkwingsmt** [comment]
> Does this actually help with solving #106552 in all cases? Let's imagine a situation where the screen is split in two halves. One half is covered by a DragGestureRecognizer, the other half isn't. Two finger (F1 and F2) touch the screen and move around, one in each half. F1 moves around in the half covered by the DragGestureRecognizer, which tracks its every move. F2 moves around outside of that area and is not tracked by the DragGestureRecognizer. In a single PointerData packet (`Iterable<ui.PointerData>`) we now get a move event for F1 and another move event for F2, which are processed by the `PointerEventConverter` in that order. The F2 event is marked with `endOfBatch`, the F1 event isn't. Unless I am missing something, the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it... Wouldn't this cause problems for #106552 (I haven't reviewed the latest changes to that PR closely, this is just from conceptually thinking about this)?

You're right, thanks for bringing this up. Well, I can't think of a way of solving this without a complete overhaul of our gesture recognizer API by disptaching the filtered batch in a list, because now that the gesture recognizer can receive any portion of the batch, how can the recognizer know that it has received the last of this portion without receiving an event from the next batch?

I would suggest reverting https://github.com/flutter/flutter/pull/106552 back to the previous approach, which is not beautiful but totally works (as far as I can see) and can resolve this issue.

**Comment by xiaoyanermiemie** [comment]
> > Does this actually help with solving #106552 in all cases? Let's imagine a situation where the screen is split in two halves. One half is covered by a DragGestureRecognizer, the other half isn't. Two finger (F1 and F2) touch the screen and move around, one in each half. F1 moves around in the half covered by the DragGestureRecognizer, which tracks its every move. F2 moves around outside of that area and is not tracked by the DragGestureRecognizer. In a single PointerData packet (`Iterable<ui.PointerData>`) we now get a move event for F1 and another move event for F2, which are processed by the `PointerEventConverter` in that order. The F2 event is marked with `endOfBatch`, the F1 event isn't. Unless I am missing something, the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it... Wouldn't this cause problems for #106552 (I haven't reviewed the latest changes to that PR closely, this is just from conceptually thinking about this)?
> 
> You're right, thanks for bringing this up. Well, I can't think of a way of solving this without a complete overhaul of our gesture recognizer API by disptaching the filtered batch in a list, because now that the gesture recognizer can receive any portion of the batch, how can the recognizer know that it has received the last of this portion without receiving an event from the next batch?
> 
> I would suggest reverting #106552 back to the previous approach, which is not beautiful but totally works (as far as I can see).

@dkwingsmt 

So, the best way is what I said "make every pointer dispatch the event even some pointer not move". Android did this, but iOS not. 

I know why only moved pointer dispatches on iOS but not on Android. https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#:~:text=%2D%20(void)dispatchTouches,(UIEvent*)event%20%7B In the dispatchTouches function, it should use the event.alltouches.

If we do this, we will not use `endOfBatch`.
But the behavior of " the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it." may be still existing.  I will check this performance this week. But  I think the `endOfBatch` is not the reason for this behavior. Even we use UIEvent.allTouches, this behaivor do not change, unless we modify the algorithem in monodrag.dart.

All above, the `UITouch.allTouches` can replace `endOfBatch`, and it is more clearer (avoid all event adding an attributes). But this replacement can not fix the problem (or we can refer the iOS's behavior).

**Comment by dkwingsmt** [comment]
I see, so you mean, if we switch to using `allTouches`, then the counting approach will be much more reliable. I agree. @goderbauer What do you think?

**Comment by xiaoyanermiemie** [comment]
> I see, so you mean, if we switch to using `allTouches`, then the counting approach will be much more reliable. I agree. @goderbauer What do you think?

@goderbauer  @dkwingsmt 

After I tested, the behavior " the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it." exists indeed.

But my fix code can handle it. Because I both use `endOfBatch`, and the count of `PointerDownEvent'. Which means that, if there are 2 pointers in `Recognizer A` and 3 pointers in `Recognizer B`. the `endOfBatch` is false in A, and true in B, but Recognizer A records 2 pointers and B records 3 pointers. So We can use this information.

But I still suggest we can use `allTouches` than `enfOfBatch`.

**Comment by goderbauer** [comment]
> I see, so you mean, if we switch to using allTouches, then the counting approach will be much more reliable. I agree. @goderbauer What do you think?

Sorry, I don't fully follow what the new proposal is.

**Comment by dkwingsmt** [comment]
> > I see, so you mean, if we switch to using `allTouches`, then the counting approach will be much more reliable. I agree. @goderbauer What do you think?
> 
> 
> 
> @goderbauer  @dkwingsmt 
> 
> 
> 
> After I tested, the behavior " the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it." exists indeed.
> 
> 
> 
> But my fix code can handle it. Because I both use `endOfBatch`, and the count of `PointerDownEvent'. Which means that, if there are 2 pointers in `Recognizer A` and 3 pointers in `Recognizer B`. the `endOfBatch` is false in A, and true in B, but Recognizer A records 2 pointers and B records 3 pointers. So We can use this information.
> 
> 
> 
> But I still suggest we can use `allTouches` than `enfOfBatch`.
> 
> 

Yeah. Can you try allTouches and without endOfBatch?

**Comment by xiaoyanermiemie** [comment]
> > > I see, so you mean, if we switch to using `allTouches`, then the counting approach will be much more reliable. I agree. @goderbauer What do you think?
> > 
> > 
> > @goderbauer  @dkwingsmt
> > After I tested, the behavior " the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it." exists indeed.
> > But my fix code can handle it. Because I both use `endOfBatch`, and the count of `PointerDownEvent'. Which means that, if there are 2 pointers in `Recognizer A`and 3 pointers in`Recognizer B`. the `endOfBatch` is false in A, and true in B, but Recognizer A records 2 pointers and B records 3 pointers. So We can use this information.
> > But I still suggest we can use `allTouches` than `enfOfBatch`.
> 
> Yeah. Can you try allTouches and without endOfBatch?

I can not. I do not know how to compile the flutter/engine☹️.

**Comment by xiaoyanermiemie** [comment]
@dkwingsmt 

I tested that changing `touches` to `allTouches`, but it causes more exceptions. there are some logics to match `touches`. For example the `pointer_data_packet_converter.cc` holds the `state_` map which is influenced by touches. If we modify `touches` to `allTouches` only, it will cause bugs. And if we modify all relative logics to match `touches` to `allTouches`. It is expensive. So I suggest we use 'endOfBatch' and my fix. This works well. What is your opinion?

**Comment by dkwingsmt** [comment]
> @dkwingsmt 
> 
> 
> 
> I tested that changing `touches` to `allTouches`, but it causes more exceptions. there are some logics to match `touches`. For example the `pointer_data_packet_converter.cc` holds the `state_` map which is influenced by touches. If we modify `touches` to `allTouches` only, it will cause bugs. And if we modify all relative logics to match `touches` to `allTouches`. It is expensive. So I suggest we use 'endOfBatch' and my fix. This works well. What is your opinion?

Anything that relies on endOfBatch is easily broken when the gesture area does not cover the entire screen. I don't think it's an approach worth discussing any more.

Sorry I was busy with other issues lately but I'll experiment on changing to allTouches soon.

**Comment by xiaoyanermiemie** [comment]
> > @dkwingsmt
> > I tested that changing `touches` to `allTouches`, but it causes more exceptions. there are some logics to match `touches`. For example the `pointer_data_packet_converter.cc` holds the `state_` map which is influenced by touches. If we modify `touches` to `allTouches` only, it will cause bugs. And if we modify all relative logics to match `touches` to `allTouches`. It is expensive. So I suggest we use 'endOfBatch' and my fix. This works well. What is your opinion?
> 
> Anything that relies on endOfBatch is easily broken when the gesture area does not cover the entire screen. I don't think it's an approach worth discussing any more.
> 
> Sorry I was busy with other issues lately but I'll experiment on changing to allTouches soon.

Oh, I forget that broken. Thank you for your expriment.  I will try to change it too.

**Comment by goderbauer** [comment]
(triage) @dkwingsmt @xiaoyanermiemie What's the status on this one? Should this be closed or do you still plan to work in this particular PR in the near future?

**Comment by xiaoyanermiemie** [comment]
> (triage) @dkwingsmt @xiaoyanermiemie What's the status on this one? Should this be closed or do you still plan to work in this particular PR in the near future?

I still plan

**Comment by xiaoyanermiemie** [comment]
> Does this actually help with solving #106552 in all cases? Let's imagine a situation where the screen is split in two halves. One half is covered by a DragGestureRecognizer, the other half isn't. Two finger (F1 and F2) touch the screen and move around, one in each half. F1 moves around in the half covered by the DragGestureRecognizer, which tracks its every move. F2 moves around outside of that area and is not tracked by the DragGestureRecognizer. In a single PointerData packet (`Iterable<ui.PointerData>`) we now get a move event for F1 and another move event for F2, which are processed by the `PointerEventConverter` in that order. The F2 event is marked with `endOfBatch`, the F1 event isn't. Unless I am missing something, the DragGestureRecognizer (which only tracks F1) will never know when the batch ended in this situation because the F2 event is never routed to it... Wouldn't this cause problems for #106552 (I haven't reviewed the latest changes to that PR closely, this is just from conceptually thinking about this)?

@dkwingsmt  @goderbauer 
In this situation, there is a bug I found at the master branch of `flutter/engine`. The reproduct approach is : Two fingers on two ListViews respectively and up them together.

Here is the exception stack (tips: you can try some times, remember together):
image


This bug block me, can you fix it firstly?

**Comment by xiaoyanermiemie** [comment]
you can refer to this demo: two ListViews.


wecom-temp-f6f6991c8f13883224caf5df96532fef

**Comment by goderbauer** [comment]
> In this situation, there is a bug I found at the master branch of flutter/engine. The reproduct approach is : Two fingers on two ListViews respectively and up them together.

Can you file a bug for this with some code to reproduce it? Thanks.

**Comment by xiaoyanermiemie** [comment]
@dkwingsmt  @goderbauer  Please see the discussion in [flutter/engine](https://github.com/flutter/engine/pull/36948)

**Comment by xiaoyanermiemie** [comment]
> 

Sorry, there is no bug. I had a mistake.

**Comment by xiaoyanermiemie** [comment]
> > In this situation, there is a bug I found at the master branch of flutter/engine. The reproduct approach is : Two fingers on two ListViews respectively and up them together.
> 
> Can you file a bug for this with some code to reproduce it? Thanks.

Sorry, there is no bug. I had a mistake.

**Comment by dkwingsmt** [comment]
@xiaoyanermiemie I tried switching to `allTouches` at https://github.com/flutter/engine/blob/206ef0f083675072844331b42b06b69429982277/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L1119 :

```diff
 - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
-  [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
+  [self dispatchTouches:event.allTouches pointerDataChangeOverride:nullptr event:event];
 }
```

(I know you talked about changing `dispatchTouches` instead but I think this is a better place.) But I didn't see any tests broken with this change. Did I make the wrong change?

**Comment by xiaoyanermiemie** [comment]
> @xiaoyanermiemie I tried switching to `allTouches` at https://github.com/flutter/engine/blob/206ef0f083675072844331b42b06b69429982277/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L1119 :
> 
> ```diff
>  - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
> -  [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
> +  [self dispatchTouches:event.allTouches pointerDataChangeOverride:nullptr event:event];
>  }
> ```
> 
> And all similar methods below it.
> 
> (I know you talked about changing `dispatchTouches` instead but I think this is a better place.) But I didn't see any tests broken with this change. Did I make the wrong change?

@dkwingsmt 
Maybe tests do not consider this situation (`down` and `move` event together in `event.allTouches`). You can use two fingers to move, and it easily to reproduct. (Remember together down and fast move :)

**Comment by goderbauer** [comment]
(Triage): I am going to close this PR as it is getting stale. I would suggest filing an issue and discussing options there.

**Pull request closed without merging**